plugin/pkg/auth/authorizer/node/graph.go

```diff
@@ -393,6 +396,20 @@ func (g *Graph) AddPod(pod *corev1.Pod) {
 			g.addEdgeToDestinationIndex_locked(e)
 		}
 	}
+
+	for _, podResourceClaim := range pod.Spec.ResourceClaims {
+		claimName, _, err := resourceclaim.Name(pod, &podResourceClaim)
```

Comment:
it looks like this pulls from `.status.resourceClaimStatuses` ... does that get populated/updated before or after a pod is bound to a node? If so, this logic that assumes all the references in the pod are immutable after node binding needs to change:

https://github.com/kubernetes/kubernetes/blob/79c02ceb73f4d64f52a9d4c46785e4c7497493d9/plugin/pkg/auth/authorizer/node/graph_populator.go#L73-L90

Reply:
> does that get populated/updated before or after a pod is bound to a node?

Before. The scheduler will not schedule pods where claims are missing.

> all the references in the pod are immutable after node binding

They are, because the claims need to be allocated and reserved before scheduling, which implies that they have a finalizer and cannot disappear underneath the pod.

This is the normal flow. If a pod gets created with `spec.nodeName` already set, then kubelet may try to run containers before the claims are ready. In that case it will encounter various error conditions (claim not created, claim not allocated or not reserved, or permission to read it not granted yet) and retry until those problems are fixed. There is work under way to do this in the resource claim controller. Nontheless this obviously should better be avoided, but the authorizer changes should also be valid in such a scenario.

Reply:
> the authorizer changes should also be valid in such a scenario

would they? in the current state of this PR, once the pod is processed with nodeName set, no updates to the pod (including status updates that would make generated resourceclaim objects resolveable) would be processed by the node authorizer (graphPopulator#updatePod short circuits if the update didn't change the pod UID or nodeName)

Reply:
> graphPopulator#updatePod short circuits if the update didn't change the pod UID or nodeName)

Hrm. That's a problem. Is extending that check to update also when `status.resourceClaimStatuses` changes okay?

Reply:
The E2E test in https://github.com/kubernetes/kubernetes/pull/118209 triggers this problem: kubelet keeps getting permission errors even after the status is updated.

Reply:
> Is extending that check to update also when status.resourceClaimStatuses changes okay?

Doing another update when a meaningful change to resourceClaimStatuses happens is ok... ideally that check inside `updatePod` would be very cheap (since we want to ignore almost all pod updates to avoid churn in the authorizer graph). Would checking `len(pod.Status.ResourceClaimStatuses) == len(oldPod.Status.ResourceClaimStatuses)` or something similarly cheap catch the updates the controller would make?

Reply:
... and this change fixes it:

```patch
diff --git a/plugin/pkg/auth/authorizer/node/graph_populator.go b/plugin/pkg/auth/authorizer/node/graph_populator.go
index aff1f80c63c..2bed2ab84fa 100644
--- a/plugin/pkg/auth/authorizer/node/graph_populator.go
+++ b/plugin/pkg/auth/authorizer/node/graph_populator.go
@@ -23,6 +23,7 @@ import (
 
        corev1 "k8s.io/api/core/v1"
        storagev1 "k8s.io/api/storage/v1"
+       apiequality "k8s.io/apimachinery/pkg/api/equality"
        "k8s.io/apimachinery/pkg/util/wait"
        corev1informers "k8s.io/client-go/informers/core/v1"
        storageinformers "k8s.io/client-go/informers/storage/v1"
@@ -78,8 +79,9 @@ func (g *graphPopulator) updatePod(oldObj, obj interface{}) {
                return
        }
        if oldPod, ok := oldObj.(*corev1.Pod); ok && oldPod != nil {
-               if (pod.Spec.NodeName == oldPod.Spec.NodeName) && (pod.UID == oldPod.UID) {
-                       // Node and uid are unchanged, all object references in the pod spec are immutable
+               if (pod.Spec.NodeName == oldPod.Spec.NodeName) && (pod.UID == oldPod.UID) &&
+                       apiequality.Semantic.DeepEqual(oldPod.Status.ResourceClaimStatuses, pod.Status.ResourceClaimStatuses) {
+                       // Node and uid are unchanged, all object references in the pod spec are immutable respectively unmodified (claim statuses).
                        klog.V(5).Infof("updatePod %s/%s, node unchanged", pod.Namespace, pod.Name)
                        return
                }
```

I am not doing feature gate testing here, based on the recent discussion that if these feature-gated fields happen to be set, components consuming such a pod should support the feature.

`graphPopulator` doesn't seem to have unit tests. Do I need to add something or is it good enough to know that there will be E2E testing which covers this aspect?

Reply:
See https://github.com/kubernetes/kubernetes/compare/571a08aa844bbc8e0d0f44f1e2b87a72b0d4e510..53127ee125a5d42db0517983aad9d6b2efecc129

Reply:
deep equality on a list of nested structs was what I was trying to avoid having to do on every update of every pod throughout the entire pod lifecycle